File: backend/src/resume_orchestrator/steps/retrieve.py
from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path

from ..schemas.job_analysis import JobAnalysis
from ..schemas.retrieval_candidates import (
    RetrievalCandidate,
    RetrievalCandidates,
    RetrievalQuery,
)
from ..sources.index_sources import index_sources
from ..sources.select_sources import filter_candidates
from ..util.fs_util import ensure_dir
from ..util.time_util import now_iso


@dataclass(slots=True)
class RetrieveStepResult:
    retrieval_candidates_path: Path
    candidate_count: int


async def retrieve_step(*, repo_root: str | Path, run_dir: str | Path) -> RetrieveStepResult:
    repo_root_p = Path(repo_root)
    run_dir_p = Path(run_dir)
    inputs_dir = ensure_dir(run_dir_p / "inputs")

    job_analysis_path = inputs_dir / "job_analysis.json"
    analysis = JobAnalysis.model_validate(
        json.loads(job_analysis_path.read_text(encoding="utf-8"))
    )

    sources_index = index_sources(repo_root_p)

    groups = analysis.tagsForGraphQL.groups if analysis.tagsForGraphQL.groups else None
    tags = analysis.tagsForGraphQL.stack if analysis.tagsForGraphQL.stack else None

    primary = filter_candidates(
        sources_index.docs,
        kinds=analysis.tagsForGraphQL.kinds,
        groups=groups,
        tags=tags,
    )

    filtered_docs = primary.docs
    matched_by_id = primary.matched_by_id
    fallback = False
    if not filtered_docs:
        secondary = filter_candidates(sources_index.docs, kinds=analysis.tagsForGraphQL.kinds)
        filtered_docs = secondary.docs
        matched_by_id = secondary.matched_by_id
        fallback = True

    candidates = RetrievalCandidates(
        schemaVersion=1,
        generatedAt=now_iso(),
        query=RetrievalQuery(
            operation="Retrieve",
            variables={
                "filter": {
                    "kinds": list(analysis.tagsForGraphQL.kinds),
                    "groups": list(groups) if groups else None,
                    "tags": list(tags) if tags else None,
                }
            },
        ),
        candidates=[
            RetrievalCandidate(
                id=doc.id,
                path=doc.path,
                kind=doc.kind,  # type: ignore[arg-type]
                group=doc.group,
                title=doc.title,
                tags=doc.tags,
                matchedFilters=["fallback:kind-only"]
                if fallback
                else matched_by_id.get(doc.id, []),
            )
            for doc in filtered_docs
        ],
    )

    retrieval_candidates_path = inputs_dir / "retrieval_candidates.json"
    retrieval_candidates_path.write_text(
        json.dumps(candidates.model_dump(), indent=2) + "\n", encoding="utf-8"
    )

    return RetrieveStepResult(
        retrieval_candidates_path=retrieval_candidates_path,
        candidate_count=len(candidates.candidates),
    )

File: backend/src/steps/retrieve.ts
import fs from "node:fs/promises";
import path from "node:path";
import { JobAnalysisSchema } from "../schemas/jobAnalysis.js";
import { RetrievalCandidates, RetrievalCandidatesSchema } from "../schemas/retrievalCandidates.js";
import { indexSources } from "../sources/indexSources.js";
import { runRetrieveQuery } from "../graph/queries.js";
import type { SourceFilter } from "../graph/schema.js";
import { ensureDir } from "../util/fsUtil.js";
import { nowIso } from "../util/timeUtil.js";

export type RetrieveStepOpts = {
  repoRoot: string;
  runDir: string;
};

export type RetrieveStepResult = {
  retrievalCandidatesPath: string;
  candidateCount: number;
};

export async function retrieveStep(opts: RetrieveStepOpts): Promise<RetrieveStepResult> {
  const inputsDir = path.join(opts.runDir, "inputs");
  await ensureDir(inputsDir);

  const jobAnalysisPath = path.join(inputsDir, "job_analysis.json");
  const raw = await fs.readFile(jobAnalysisPath, "utf8");
  const analysis = JobAnalysisSchema.parse(JSON.parse(raw));

  const index = await indexSources({ root: opts.repoRoot });

  const filter: SourceFilter = {
    kinds: analysis.tagsForGraphQL.kinds,
    groups: analysis.tagsForGraphQL.groups.length > 0 ? analysis.tagsForGraphQL.groups : undefined,
    tags: analysis.tagsForGraphQL.stack.length > 0 ? analysis.tagsForGraphQL.stack : undefined
  };

  const { result, matchedFiltersById, variables, operation } = await runRetrieveQuery({ index, filter });

  if (result.errors && result.errors.length > 0) {
    throw new Error(`GraphQL retrieve errors: ${result.errors.map((e) => e.message).join("; ")}`);
  }

  let sources = result.data?.sources ?? [];

  // If tag/group filters yielded nothing, fall back to all in-kind sources so
  // ranking still has something to score. The candidate audit records the
  // fallback so reviewers can see retrieval emptied out.
  let fallback = false;
  if (sources.length === 0) {
    const fallbackResult = await runRetrieveQuery({
      index,
      filter: { kinds: analysis.tagsForGraphQL.kinds }
    });
    sources = fallbackResult.result.data?.sources ?? [];
    fallback = true;
  }

  const candidates: RetrievalCandidates = RetrievalCandidatesSchema.parse({
    schemaVersion: 1,
    generatedAt: nowIso(),
    query: {
      operation,
      variables: variables as Record<string, unknown>
    },
    candidates: sources.map((s) => ({
      id: s.id,
      path: s.path,
      kind: s.kind,
      group: s.group,
      title: s.title,
      tags: s.tags,
      matchedFilters: fallback ? ["fallback:kind-only"] : matchedFiltersById.get(s.id) ?? []
    }))
  });

  const retrievalCandidatesPath = path.join(inputsDir, "retrieval_candidates.json");
  await fs.writeFile(retrievalCandidatesPath, JSON.stringify(candidates, null, 2) + "\n", "utf8");

  return { retrievalCandidatesPath, candidateCount: candidates.candidates.length };
}
